gowin/arch.cc

```diff
@@ -0,0 +1,1248 @@
+    }
+
+    package = nullptr;
+    for (unsigned int i = 0; i < variant->num_packages; i++) {
+        auto pkg = &variant->packages[i];
+        // std::cout << IdString(pkg->name_id).str(this) << std::endl;
+        if (IdString(pkg->name_id) == id(args.package)) {
+            package = pkg;
+            break;
+        }
+        // for (int j=0; j < pkg->num_pins; j++) {
+        //     auto pin = pkg->pins[j];
+        //     std::cout << IdString(pin.src_id).str(this) << " " << IdString(pin.dest_id).str(this) << std::endl;
+        // }
+    }
+    if (package == nullptr) {
+        log_error("Unsuported package '%s'.\n", args.package.c_str());
+    }
+    // setup db
+    char buf[32];
+    for (int i = 0; i < db->rows * db->cols; i++) {
+        int row = i / db->cols;
+        int col = i % db->cols;
+        const TilePOD *tile = db->grid[i].get();
+        // setup wires
+        const PairPOD *pips[2] = {tile->pips.get(), tile->clock_pips.get()};
+        unsigned int num_pips[2] = {tile->num_pips, tile->num_clock_pips};
+        for (int p = 0; p < 2; p++) {
+            for (unsigned int j = 0; j < num_pips[p]; j++) {
+                const PairPOD pip = pips[p][j];
+                int destrow = row;
+                int destcol = col;
+                IdString destid = pip.dest_id;
+                IdString gdestname = wireToGlobal(destrow, destcol, db, destid);
+                if (wires.count(gdestname) == 0)
+                    addWire(gdestname, pip.dest_id, destcol, destrow);
+                int srcrow = row;
+                int srccol = col;
+                IdString srcid = pip.src_id;
+                IdString gsrcname = wireToGlobal(srcrow, srccol, db, srcid);
+                if (wires.count(gsrcname) == 0)
+                    addWire(gsrcname, pip.src_id, srccol, srcrow);
+            }
+        }
+        for (unsigned int j = 0; j < tile->num_bels; j++) {
+            const BelsPOD *bel = &tile->bels[j];
+            IdString belname;
+            IdString wirename;
+            IdString portname;
+            int z = 0;
+            bool dff = true;
+            switch (static_cast<ConstIds>(bel->type_id)) {
+            // fall through the ++
+            case ID_LUT7:
+                z++;
+                dff = false;
+            case ID_LUT6:
+                z++;
+                dff = false;
+            case ID_LUT5:
+                z++;
+            case ID_LUT4:
+                z++;
+            case ID_LUT3:
+                z++;
+            case ID_LUT2:
+                z++;
+            case ID_LUT1:
+                z++;
+            case ID_LUT0:
+                // common LUT+DFF code
+                snprintf(buf, 32, "R%dC%d_SLICE%d", row + 1, col + 1, z);
+                belname = id(buf);
+                addBel(belname, id_SLICE, Loc(col, row, z), false);
+                snprintf(buf, 32, "R%dC%d_F%d", row + 1, col + 1, z);
+                wirename = id(buf);
```

Comment:
is there any need for the wirename temporary variable - wouldn't doing `id(buf)` as an argument to `addBelOutput` etc be neater?

Reply:
I also think you could have a list of pin names and do this in a loop

Reply:
They are a mix of inputs and outputs, so I think it would be less clear in the end.